File: p42895.cpp
#include<bits/stdc++.h>
using namespace std;
set<int> dp[10];
int N;
set<int> go(int k){
    set<int> &ret = dp[k];
    if(!ret.empty()) return dp[k];
    int num = 0;
    for(int i = 0; i < k; i++) num = num * 10 + N;
    ret.insert(num);
    for(int i = 1; i < k; i++){
        set<int> s1 = go(i);
        set<int> s2 = go(k - i);
        for(int n1 : s1){
            for(int n2 : s2){
                ret.insert(n1 + n2);
                ret.insert(n1 - n2);
                ret.insert(n1 * n2);
                if(n2 != 0) ret.insert(n1 / n2);
            }
        }
    }
    return ret;
}
int solution(int _N, int number) {
    N = _N;
    for(int i = 1; i <= 8; i++){
        dp[i] = go(i);
        if(dp[i].find(number) != dp[i].end()) return i;
    }
    return -1;
}
int main(){
    int N, number;
    cin >> N >> number;
    cout << solution(N, number) << "\n"; 
}

File: p42895.java
public class p42895 {
    public static int ret = -1;
    public int solution(int N, int number) {
        dfs(N, number, 0, 0);
        return ret;
    }

    public static void dfs(int N, int number, int depth, int val){
        if(depth > 8) return;
        if(val == number){
            if(ret == -1) ret = depth;
            else ret = Math.min(ret, depth);
            return;
        }

        int k = N;
        for(int i = 1; i <= 8 - depth; i++){
            dfs(N, number, depth + i, val + k);
            dfs(N, number, depth + i, val - k);
            dfs(N, number, depth + i, val / k);
            dfs(N, number, depth + i, val * k);   
            k = k * 10 + N;
        }
    }
}
